### PR TITLE
Add high-level documentation for protocol developers

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Updated with info on connections and RoCs.

**Comment by gavin-norman-sociomantic** [comment]
The sections on the basic protocol and request / connection handling are now complete. I suggest reviewing and merging those, and writing further documentation in another PR.

File: src/swarm/neo/protocol/connect/ConnectProtocol.d [hunk 1]
```diff
@@ -205,15 +205,14 @@ class ConnectProtocol: ISelectClient
...
         Params:
-            client_timestamp = client time stamp
-            node_timestamp   = client time stamp
+            client_protocol = client protocol version
+            node_protocol   = client protocol version
 
-        Returns:
-            this.protocol_error_.
+        Throws:
+            if there is a protocol version mismatch
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/protocol/connect/ConnectProtocol.d]: I think it's fair enough to specify in addition to this that it's throwing a protocol error.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/protocol/connect/ConnectProtocol.d]: The exception type you mean?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/protocol/connect/ConnectProtocol.d]: Yes, sorry I should have written: `ProtocolError`

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/protocol/connect/ConnectProtocol.d]: Updated.

File: src/swarm/README_protocol_neo.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  connection.
+* An event dispatcher (the nested class `EventDispatcher`) instance that
+  provides the following functionality:
+  - A method to suspend the RoC fiber and wait for one of a set of specified
+    events. Events that can be waited on are:
+    1. Receiving a payload for this request.
+    2. Sending a payload.
+    3. The fiber being resumed after it was yielded and an epoll event-loop
+       cycle occurred.
+    4. The fiber being resumed with a non-negative code.
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/README_protocol_neo.md]: It might be worth mentioning that the fiber is here _explicitly_ resumed.

Comment (gavin-norman-sociomantic) [reply src/swarm/README_protocol_neo.md]: Hmm, what do you mean by "explicitly"? Is it possible for a fiber to be non-explicitly resumed?

Comment (nemanja-boric-sociomantic) [reply src/swarm/README_protocol_neo.md]: By explicitly calling `resumeFiber` method. AFAIR, all other cases resume fiber with the negative code, right? I think it's useful to say that the non-negative codes are reserved for explicit `resumeFiber` calls.

Comment (nemanja-boric-sociomantic) [reply src/swarm/README_protocol_neo.md]: Ping :-)

Comment (gavin-norman-sociomantic) [reply src/swarm/README_protocol_neo.md]: Updated.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

File: src/swarm/README_protocol_neo.md (hunk 2 above)
Comment (nemanja-boric-sociomantic) [reply src/swarm/README_protocol_neo.md]: Thanks!

**Pull request merged**